File: tools/Python/t4-Python-api/src/t4login/definitions/chartdata/chart_data_type.py
"""Port of `com.t4login.definitions.chartdata.ChartDataType`.

Ported as a plain class (not IntEnum) to preserve the Java `@AsEnum` runtime
extensibility: `get(value)` dynamically registers unknown values with a logged
warning, and `values()` returns the live list.

Represents the time-aggregation level of chart data:
- Tick (0)   — individual trades, no time aggregation.
- Second (1) — bars aggregated to 1-second boundaries.
- Minute (2) — bars aggregated to 1-minute boundaries.
- Hour (3)   — bars aggregated to 1-hour boundaries.
- Day (4)    — daily bars (one per trade date).
- TPO (5)    — Time-Price Opportunity (market profile) data.
- TickChange (6) — price-change records (tick-level, no volume).

The stream readers use these values to determine how to truncate bar start
times (via ``get_bar_start_time``) and to select which decode path to follow.
"""

from __future__ import annotations

import logging

_log = logging.getLogger(__name__)

_map: dict[int, ChartDataType] = {}
_values: list[ChartDataType] = []


class ChartDataType:
    """Chart data aggregation type (mirrors Java's `@AsEnum` pattern)."""

    __slots__ = ("_name", "_value")

    _name: str
    _value: int

    def __init__(self, value: int, name: str) -> None:
        self._value = value
        self._name = name

    @property
    def value(self) -> int:
        return self._value

    @property
    def name(self) -> str:
        return self._name

    def get_value(self) -> int:
        """Java-style getter alias."""
        return self._value

    @staticmethod
    def get(value: int) -> ChartDataType:
        """Lookup by int value; dynamically registers unknown values (mirrors Java)."""
        val = _map.get(value)
        if val is None:
            _log.warning(
                "ChartDataType.get(): non-existent value %d created and added without name.",
                value,
            )
            val = ChartDataType(value, str(value))
            _map[value] = val
            _values.append(val)
        return val

    @staticmethod
    def values() -> list[ChartDataType]:
        """Returns the live list of all known instances."""
        return list(_values)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ChartDataType):
            return NotImplemented
        return self._value == other._value

    def __hash__(self) -> int:
        return hash(self._value)

    def __repr__(self) -> str:
        return f"ChartDataType({self._value}, {self._name!r})"

    def __str__(self) -> str:
        return self._name


def _register(value: int, name: str) -> ChartDataType:
    inst = ChartDataType(value, name)
    _map[value] = inst
    _values.append(inst)
    return inst


# --- Well-known instances (mirrors Java static block) -------------------------
Tick: ChartDataType = _register(0, "Tick")
Second: ChartDataType = _register(1, "Second")
Minute: ChartDataType = _register(2, "Minute")
Hour: ChartDataType = _register(3, "Hour")
Day: ChartDataType = _register(4, "Day")
TPO: ChartDataType = _register(5, "TPO")
TickChange: ChartDataType = _register(6, "TickChange")

File: tools/Java/t4-java-decoder/src/main/java/com/t4login/definitions/chartdata/ChartDataType.java
package com.t4login.definitions.chartdata;

import com.t4login.AsEnum;
import com.t4login.Constant;
import com.t4login.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Chart data aggregation type.
 */
@AsEnum
public class ChartDataType {

    private static final String TAG = "ChartDataType";

    // @formatter:off
    @Constant(0)
    public static final ChartDataType Tick;
    @Constant(1)
    public static final ChartDataType Second;
    @Constant(2)
    public static final ChartDataType Minute;
    @Constant(3)
    public static final ChartDataType Hour;
    @Constant(4)
    public static final ChartDataType Day;
    @Constant(5)
    public static final ChartDataType TPO;
    @Constant(6)
    public static final ChartDataType TickChange;
    // @formatter:on

    private final int value;
    private final String name;
    private static Map<Integer, ChartDataType> map = new HashMap<>();
    private static ChartDataType[] mValues;

    static {
        List<ChartDataType> values = new ArrayList<>();

        Tick = new ChartDataType(0, "Tick");
        values.add(Tick);
        Second = new ChartDataType(1, "Second");
        values.add(Second);
        Minute = new ChartDataType(2, "Minute");
        values.add(Minute);
        Hour = new ChartDataType(3, "Hour");
        values.add(Hour);
        Day = new ChartDataType(4, "Day");
        values.add(Day);
        TPO = new ChartDataType(5, "TPO");
        values.add(TPO);
        TickChange = new ChartDataType(6, "TickChange");
        values.add(TickChange);

        for (ChartDataType t : values) {
            map.put(t.getValue(), t);
        }

        mValues = new ChartDataType[values.size()];
        values.toArray(mValues);
    }

    ChartDataType(int value, String name) {
        this.value = value;
        this.name = name;
    }

    public int getValue() {
        return this.value;
    }

    public static ChartDataType get(int value) {
        ChartDataType val = map.get(value);

        if (val == null) {
            Log.e(TAG, "get(), Non-existent value " + Integer.toString(value) + " created and added without name." + "");
            val = new ChartDataType(value, Integer.toString(value));
            map.put(value, val);

            List<ChartDataType> values = new ArrayList<>();
            Collections.addAll(values, mValues);
            values.add(val);

            mValues = new ChartDataType[values.size()];
            values.toArray(mValues);
        }

        return val;
    }

    public ChartDataType[] values() {
        return mValues;
    }

    @Override
    public String toString() {
        return this.name;
    }
}
